### ISSUE TITLE
Narrowing to a protocol loses information about self type

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#6434]
```python
import dataclasses
from dataclasses import dataclass, replace
from typing import Any, ClassVar, Generic, Protocol, Self, TypeVar


class DataclassInstance(Protocol):
    __dataclass_fields__: ClassVar[dict[str, dataclasses.Field[Any]]]


T = TypeVar('T')
@dataclass
class X(Generic[T]):
    a: int

    def f(self) -> Self:
        reveal_type(self)  # Self@X[T@X]
        assert isinstance(self, DataclassInstance)
        reveal_type(self)  # X[T@X]
        return replace(self, a=1)  # Error, can't return X.
```

**Comment by erictraut** [comment]
I've modified the `isinstance` narrowing logic to support this case. This change has some tradeoffs, and it could result in other false positives. The `mypy_primer` regression tests seem to indicate that the impact should be relatively limited, but if it turns out to be more impactful than I fear, I may need to revisit this case.

If and when we add intersection types to the type system, we'll be able to model this case in a way that doesn't involve such tradeoffs.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.336, which I just published. It will also be included in a future release of pylance.
